File: hooks/freshness-nudge.py
#!/usr/bin/env python3
"""SessionStart nudge when a project's instruction files have gone stale.

Cheap and throttled: emits at most one gentle, read-only note into context, and
only when the set of stale references has changed or it has been a while since
the last note. Never edits the user's files. Never breaks session start.

Standard library only. Python 3.10+.
"""

from __future__ import annotations

import json
import os
import sys
from datetime import date
from pathlib import Path

# Resolve scripts/ as an import location so we can reuse drift + its deps.
sys.path.insert(0, str(Path(__file__).resolve().parent.parent / "scripts"))
import drift as drift_mod  # noqa: E402

THROTTLE_DAYS = 14


def project_dir() -> Path:
    return Path(os.environ.get("CLAUDE_PROJECT_DIR") or os.getcwd())


def main() -> None:
    root = project_dir()
    result = drift_mod.scan(str(root))
    if not result["stale_files"]:
        return

    marker = root / ".hestia" / "freshness-nudge.json"
    prev: dict = {}
    try:
        prev = json.loads(marker.read_text(encoding="utf-8"))
    except (OSError, ValueError):
        prev = {}

    same_sig = bool(prev.get("signature")) and prev.get("signature") == result["signature"]
    recent = False
    if same_sig and prev.get("date"):
        try:
            recent = (date.today() - date.fromisoformat(prev["date"])).days < THROTTLE_DAYS
        except ValueError:
            recent = False
    # Derived staleness (read-time label from the last checkup's cheap signal; no
    # grade is stored). A setup that has aged to "stale" warrants a reminder even
    # when the broken-ref signature itself hasn't changed — so it overrides the
    # throttle. Everything else keeps the existing behavior.
    setup_stale = (result.get("staleness", {}) or {}).get("label") == "stale"
    if same_sig and recent and not setup_stale:
        return  # already nudged for this exact staleness, recently

    try:
        marker.parent.mkdir(parents=True, exist_ok=True)
        marker.write_text(
            json.dumps({"signature": result["signature"], "date": date.today().isoformat()}),
            encoding="utf-8",
        )
    except OSError:
        pass

    files = ", ".join(s["path"] for s in result["stale_files"][:5])
    n = result["total_broken"]
    msg = (
        f"Hestia freshness check: {n} reference(s) in this project's instruction files "
        f"point to missing paths (in {files}). At a natural moment, gently let the user know "
        f"and offer to run /hestia:checkup or /hestia:freshness. Mention this once; do not nag."
    )
    try:
        sys.stdout.buffer.write(msg.encode("utf-8"))
    except OSError:
        pass


if __name__ == "__main__":
    try:
        main()
    except Exception:
        # A freshness nudge must never break session start.
        sys.exit(0)

File: hooks/freshness-nudge.js
"use strict";

const fs = require("fs");
const path = require("path");

const { scan } = require("../scripts/drift");

const THROTTLE_DAYS = 14;

function projectDir() {
  return path.resolve(process.env.CLAUDE_PROJECT_DIR || process.cwd());
}

function today() {
  return new Date().toISOString().slice(0, 10);
}

function daysSince(isoDate) {
  const then = new Date(isoDate + "T00:00:00Z");
  if (Number.isNaN(then.getTime())) return null;
  const now = new Date(today() + "T00:00:00Z");
  return Math.round((now - then) / 86400000);
}

function main() {
  const root = projectDir();
  const result = scan(root);
  if (!result.stale_files.length) return;

  const marker = path.join(root, ".hestia", "freshness-nudge.json");
  let prev = {};
  try {
    prev = JSON.parse(fs.readFileSync(marker, "utf-8"));
  } catch {
    prev = {};
  }

  const sameSig = Boolean(prev.signature) && prev.signature === result.signature;
  let recent = false;
  if (sameSig && prev.date) {
    const diff = daysSince(prev.date);
    recent = diff !== null && diff < THROTTLE_DAYS;
  }
  const setupStale = (result.staleness || {}).label === "stale";
  if (sameSig && recent && !setupStale) return;

  try {
    fs.mkdirSync(path.dirname(marker), { recursive: true });
    fs.writeFileSync(
      marker,
      JSON.stringify({ signature: result.signature, date: today() }),
      "utf-8"
    );
  } catch {
    // best-effort
  }

  const files = result.stale_files
    .slice(0, 5)
    .map((s) => s.path)
    .join(", ");
  const n = result.total_broken;
  const msg =
    `Hestia: ${n} broken reference(s) in this project's instruction files (${files}) ` +
    `— Claude is working from stale paths. Surface this to the user at the first ` +
    `natural break and offer /hestia:freshness or /hestia:checkup; don't sit on it.`;
  try {
    process.stdout.write(Buffer.from(msg, "utf-8"));
  } catch {
    // stdout write failure -> mirror Python's OSError swallow
  }
}

if (require.main === module) {
  try {
    main();
  } catch {
    process.exit(0);
  }
}

module.exports = { projectDir, main };
